spring-cloud-gcp-samples/spring-cloud-gcp-pubsub-config-sample/README.adoc

```diff
@@ -0,0 +1,78 @@
+= Spring Cloud GCP Pub/Sub Configuration Management Code Sample
+
+This code sample demonstrates automated configuration management using Spring Cloud Config, Spring Cloud Bus and GCP Pub/Sub.
+The sample app consists of a configuration server and a small web application that subscribes to the server to be notified of property changes.
+
+Configuration server runs on port `8888`, and can read configuration from a local filesystem (useful for exploring the sample) or from a source control system.
+When the server is notified that new configuration is available, it fetches the updated configuration and sends it out via Spring Cloud Bus.
+Under the hood, Spring Cloud Bus uses the Spring Cloud Stream binder for GCP Pub/Sub to send a message to a topic named `springCloudBus`.
+The client uses the same binder with an anonymous subscription to receive notifications, and provides an HTTP endpoint `/message` for printing the dynamically updated message value.
+
+== Setup
+
+1. Configure your GCP project ID and credentials by following link:../../docs/src/main/asciidoc/core.adoc#project-id[these instructions].
++
+Alternatively, if you have the https://cloud.google.com/sdk/[Google Cloud SDK] installed and initialized, and are logged in with https://developers.google.com/identity/protocols/application-default-credentials[application default credentials], Spring will auto-discover those parameters for you.
+
+2. Make sure that GCP Cloud Pub/Sub API is enabled.
+
+== Running the code sample with configuration derived from local filesystem
+
+1. In separate terminal windows, start the `spring-cloud-gcp-pubsub-config-sample-server` and `spring-cloud-gcp-pubsub-config-sample-client` applications with the `$ mvn spring-boot:run` command in the same folder as the apps' `pom.xml` files.
+
+2. Visit http://localhost:8080/message in a browser; observe the default value of `none` printed on the page.
+This happens because there is no configuration available for config server yet.
+
+3. Create a temporary directory `/tmp/config`.
+Alternatively, update `spring.cloud.config.server.native.searchLocationsconfig` property in configuration server's link:spring-cloud-gcp-pubsub-config-sample-server/src/main/resources/application.properties[application.properties] with the directory you will use, and restart the config server.
+
+4. Place a file named `application.properties` with the following contents into the directory above:
++
+----
+example.message = message from the local filesystem
+----
+
+5. Notify the configuration server that new configuration is available.
+In a deployed server, this would be done through a GitHub webhook, or similar mechanism.
++
+----
+curl -X POST http://localhost:8888/monitor -d 'path=application'
+----
+
+6. Visit http://localhost:8080/message in a browser; observe the value now matches the value updated in the file system.
```

Comment:
Before this step, you can ask the user to observe the server log and the client log:
Server:
```
INFO 199601 --- [nio-8888-exec-4] o.s.cloud.bus.event.RefreshListener      : Received remote refresh request. Keys refreshed []
INFO 199601 --- [nio-8888-exec-6] o.s.c.c.s.e.NativeEnvironmentRepository  : Adding property source: file:/tmp/config/application.properties
```

Client:
```
INFO 204308 --- [sub-subscriber1] c.c.c.ConfigServicePropertySourceLocator : Fetching config from server at : http://localhost:8888
INFO 204308 --- [sub-subscriber1] c.c.c.ConfigServicePropertySourceLocator : Located environment: name=application, profiles=[default], label=null, version=null, state=null
INFO 204308 --- [sub-subscriber1] b.c.PropertySourceBootstrapConfiguration : Located property source: CompositePropertySource {name='configService', propertySources=[MapPropertySource {name='file:/tmp/config/application.properties'}]}
INFO 204308 --- [sub-subscriber1] o.s.boot.SpringApplication               : No active profile set, falling back to default profiles: default
INFO 204308 --- [sub-subscriber1] o.s.boot.SpringApplication               : Started application in 0.162 seconds (JVM running for 553.417)
INFO 204308 --- [sub-subscriber1] o.s.cloud.bus.event.RefreshListener      : Received remote refresh request. Keys refreshed [example.message]
```